segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java

```diff
@@ -106,19 +144,111 @@ public void stop() {
+        if (secureZk) {
+            return waitForSSLServerUp(address, LocalBookKeeper.CONNECTION_TIMEOUT, trustStore, keyStore,
+                    keyStorePasswordPath, trustStorePasswordPath);
+        } else {
+            return LocalBookKeeper.waitForServerUp(address, LocalBookKeeper.CONNECTION_TIMEOUT);
+        }
+    }
+
     public static boolean waitForServerUp(int zkPort) {
-        val address = LOOPBACK_ADDRESS.getHostAddress() + ":" + zkPort;
-        return LocalBookKeeper.waitForServerUp(address, LocalBookKeeper.CONNECTION_TIMEOUT);
+        return waitForServerUp(zkPort, false, "", "", "", "");
+    }
+
+    private static boolean waitForSSLServerUp(String address, long timeout, String trustStore, String keyStore,
+                                              String keyStorePasswdPath, String trustStorePasswordPath) {
+        TimeoutTimer timeoutTimer = new TimeoutTimer(Duration.ofMillis(timeout));
+        String[] split = address.split(":");
+        String host = split[0];
+        int port = Integer.parseInt(split[1]);
+
+        while (true) {
+            try {
+                SSLContext context = SSLContext.getInstance("TLS");
+                TrustManagerFactory trustManager = getTrustManager(trustStore, trustStorePasswordPath);
+                KeyManagerFactory keyFactory = getKeyManager(keyStore, keyStorePasswdPath);
+                context.init(keyFactory.getKeyManagers(), trustManager.getTrustManagers(), null);
+
+                try (Socket sock = context.getSocketFactory().createSocket(new Socket(host, port), host, port, true);
+                     OutputStream outstream = sock.getOutputStream()) {
+                    outstream.write("stat".getBytes());
+                    outstream.flush();
+
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(sock.getInputStream()));
+                    String line = reader.readLine();
+                    if (line != null && line.startsWith("Zookeeper version:")) {
+                        log.info("Server UP");
+                        return true;
+                    }
+                }
+            } catch (IOException | CertificateException | NoSuchAlgorithmException | KeyStoreException
+                    | KeyManagementException | UnrecoverableKeyException e) {
+                // ignore as this is expected
+                log.warn("server  {} not up.", address,  e);
+            }
+
+            if (!timeoutTimer.hasRemaining()) {
+                break;
+            }
+            Exceptions.handleInterrupted(() -> Thread.sleep(250));
+        }
+        return false;
+    }
+
+    private static TrustManagerFactory getTrustManager(String trustStore, String trustStorePasswordPath)
+            throws IOException, CertificateException, NoSuchAlgorithmException, KeyStoreException {
+        try (FileInputStream myKeys = new FileInputStream(trustStore)) {
+
+            // Do the same with your trust store this time
```

Comment:
what does "your trust store" mean? Same with the next line.

This looks suspiciously copied from https://stackoverflow.com/questions/24555890/using-a-custom-truststore-in-java-as-well-as-the-default-one/24561444

Reply:
Removed irrelevant comments.